Device tracker component & platform validation. No more home_range.

kellerza opened the pull request:
**Description:**
*Device tracker component & validation for platforms
*Removed home_range from device_tracker

**Example entry for `configuration.yaml` (if applicable):**

``` yaml
device_tracker:
```

If the code does not interact with devices:
- [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
- [x] Tests have been added to verify that the new code works.

kellerza committed:
Device tracker component & platform validation. No more home_range.

balloob commented on homeassistant/components/device_tracker/__init__.py:
```diff
@@ -95,30 +103,43 @@ def setup(hass, config):
+            if hasattr(platform, 'PLATFORM_SCHEMA'):
+                p_config = platform.PLATFORM_SCHEMA(p_config)
+        except vol.MultipleInvalid as ex:
+            _log_exception(ex, DOMAIN, p_config)
```
Please rename this function to remove the `_` prefix. The convention is Python that `_` prefix means that it's private to that module and should not be used.

balloob commented on homeassistant/components/device_tracker/__init__.py:
```diff
@@ -95,30 +103,43 @@ def setup(hass, config):
-    if isinstance(conf, list):
-        conf = conf[0] if len(conf) > 0 else {}
+    def _get_conf_value(key_name, default):
+        """Retrieve a value from all platforms & rasie error if conflictind."""
```
`raise`, `conflicting`

kellerza replied:
:smile: docstrings needs a spell checker!

balloob commented on homeassistant/components/device_tracker/__init__.py:
```diff
@@ -95,30 +103,43 @@ def setup(hass, config):
-    # Config can be an empty list. In that case, substitute a dict
-    if isinstance(conf, list):
-        conf = conf[0] if len(conf) > 0 else {}
+    def _get_conf_value(key_name, default):
```
I don't like this approach. It is already confusing enough that a config entry for 1 platform applies to all platforms. I feel that by limiting it to just the first platform it is at least less confusing. Eventually we should get a better way to specify device tracker not-platform-related config.

kellerza replied:
With this approach you can add it anywhere (first, last, or more than one places), but it will warn you if you use different values (we can and will only honor the first one)

It will solve issues, where you add a dev tracker in the future, but inserting it first in the list for example. Currently it will just silently ignore

balloob commented on homeassistant/components/device_tracker/__init__.py:
```diff
@@ -95,30 +103,43 @@ def setup(hass, config):
+            return res[0]
+        return default
 
+    _LOGGER.warning("CONFIG: from device_tracker.setup() %s", conf)
```
stale log

balloob commented on homeassistant/components/device_tracker/__init__.py:
```diff
@@ -95,30 +103,43 @@ def setup(hass, config):
         if platform is None:
             return
 
+        # Apply component & platform specific validation schema
```
We shouldn't have to do this. This is already done inside bootstrap.

kellerza replied:
Exactly what I thought, but it seem like after I added the CONFIG_SCHEMA to device_tracker, this is not happening any more [related to this elif?](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/bootstrap.py#L110)

balloob replied:
So then let's not define a `CONFIG_SCHEMA`?

kellerza committed:
Mock, bluetooth

balloob commented:
Your commits seem weird for this PR? It includes secrets , vsure bump

kellerza commented:
I forced the "Mock, bluetooth" commit and it seemed to have caused issues. Will separate tomorrow

balloob commented:
Bump :)

kellerza commented:
- I fixed some feedback in "Fix feedback" patch (`log_exception` and made `_component_value` clearer)
 - Or should I rather remove the `_component_value` function in line 108? (It does help for validation though :smile:)
- With the validation bit I am stuck. And if I don't do it MQTT fails since no default values added by PLATFORM_SCHEMA
 - I've added a unit test for that (check if Qos is there) if you want to look into it. For this it might be better to look at bootstrap more and eventually remove it here.

balloob commented on homeassistant/components/device_tracker/__init__.py:
```diff
@@ -95,30 +103,39 @@ def setup(hass, config):
-        conf = conf[0] if len(conf) > 0 else {}
+    def _component_value(key_name, default):
+        """Retrieve a config value & ensure all platforms agree."""
+        res = list(set([plat[key_name] for plat in conf if key_name in plat]))
```
I still don't like this. Why even bother checking the other platforms ?

kellerza committed:
Renamed _CONFIG_SCHEMA. Raise warning for #1606

robbiet480 commented:
@kellerza What's the status on this? Good to merge?

kellerza commented:
@robbiet480 Let me just add a test for the duplicate MAC / dev_id detection to at least keep coverage the same

kellerza committed:
test duplicates

robbiet480 commented:
Hey @kellerza, got an update on this? Looks like tests just need to be fixed then it's good to go?

kellerza commented:
Seems like `mock_f.assert_called_once()` was removed in Python 3.5 and `assert mock_f.call_count == 1` is the preferred way. I'll fix

kellerza committed:
Fix assert

kellerza committed:
Coverage

kellerza committed:
Typing

kellerza commented:
@robbiet480 should be ok now. Coverage up 0.4%

Teagan42 commented on tests/components/device_tracker/test_init.py:
```diff
@@ -50,38 +55,71 @@ def test_reading_broken_yaml_config(self):  # pylint: disable=no-self-use
+            device_tracker.Device(self.hass, True, True, 'your_device',
+                                  'AB:01', 'Your device', None, None, False)]
+        device_tracker.DeviceTracker(self.hass, False, True, devices)
+        print(mock_warning.call_args_list)
```
Stale print

kellerza replied:
By design, or should I rather add it to the assert string below?

Teagan42 commented on tests/components/device_tracker/test_init.py:
```diff
@@ -12,20 +12,25 @@
+    yaml_devices = None  # type: str
 
-    def setUp(self):  # pylint: disable=invalid-name
+    def setup_method(self, _):
```
Do we really need to change this?

kellerza replied:
Based on trying to get rid of pylint disables, yes.

But also not wise to follow rules blindly. Keep the disable?

Teagan42 commented on tests/components/device_tracker/test_init.py:
```diff
@@ -12,20 +12,25 @@
         self.yaml_devices = self.hass.config.path(device_tracker.YAML_DEVICES)
 
-    def tearDown(self):  # pylint: disable=invalid-name
+    def teardown_method(self, _):
```
Do we really need to change this?

kellerza committed:
T fixes

Teagan42 commented:
🐬
